src/inc/til/bitmap.h

```diff
@@ -0,0 +1,221 @@
+            }
+            else
+            {
+                // If we reached the end, set the pos because the run is empty.
+                _pos = _nextPos;
+                _run = til::rectangle{};
+            }
+        }
+    };
+
+    class bitmap
+    {
+    public:
+        using const_iterator = _bitmap_const_iterator;
+
+        bitmap() :
+            bitmap(til::size{ 0, 0 })
+        {
+        }
+
+        bitmap(til::size sz) :
+            _sz(sz),
+            _rc(sz),
+            _bits(sz.area(), false),
+            _empty(true)
+        {
+        }
+
+        const_iterator begin() const
+        {
+            return _bitmap_const_iterator(_bits, _sz, 0);
+        }
+
+        const_iterator end() const
+        {
+            return _bitmap_const_iterator(_bits, _sz, _sz.area());
+        }
+
+        void set(til::point pt)
+        {
+            THROW_HR_IF(E_INVALIDARG, !_rc.contains(pt));
+
+            _bits[_rc.index_of(pt)] = true;
+            _empty = false;
+        }
+
+        void reset(til::point pt)
+        {
+            THROW_HR_IF(E_INVALIDARG, !_rc.contains(pt));
+
+            _bits[_rc.index_of(pt)] = false;
+        }
+
+        void set(til::rectangle rc)
+        {
+            THROW_HR_IF(E_INVALIDARG, !_rc.contains(rc));
+
+            for (auto pt : rc)
+            {
+                set(pt);
+            }
+        }
+
+        void reset(til::rectangle rc)
+        {
+            THROW_HR_IF(E_INVALIDARG, !_rc.contains(rc));
+
+            for (auto pt : rc)
+            {
+                reset(pt);
+            }
+        }
+
+        void set_all()
+        {
+            // .clear() then .resize(_size(), true) throws an assert (unsupported operation)
+            // .assign(_size(), true) throws an assert (unsupported operation)
+
+            set(_rc);
+        }
+
+        void reset_all()
+        {
+            // .clear() then .resize(_size(), false) throws an assert (unsupported operation)
+            // .assign(_size(), false) throws an assert (unsupported operation)
+            reset(_rc);
+            _empty = true;
+        }
+
+        // True if we resized. False if it was the same size as before.
+        bool resize(til::size size)
+        {
+            // Don't resize if it's not different as we mark the whole thing dirty on resize.
+            if (_sz != size)
+            {
+                _sz = size;
+                _rc = til::rectangle{ size };
+                // .resize(_size(), true/false) throws an assert (unsupported operation)
+                _bits = std::vector<bool>(_sz.area(), false);
+                _empty = true;
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        constexpr bool empty() const
+        {
+            return _empty;
+        }
+
+    private:
+        bool _empty;

```

Comment:
this scares me. setting and resetting a single point = empty _state_ but not _`_empty`_

Reply:
It's gone now with the revision.